compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseConsumerIntoLoop.cpp

```diff
@@ -0,0 +1,742 @@
+  rewriter.eraseOp(clonedConsumerOp);
+
+  return SCFFuseConsumerOfSliceResult{
+      consumerOp, tileAndFuseResult->tiledOps[0], {}};
+}
+
+/// Implementation of fusing consumer of a single slice by computing the
+/// slice of the consumer in-place.
+static FailureOr<SCFFuseConsumerOfSliceResult> tileAndFuseConsumerOfSlice(
+    RewriterBase &rewriter, Operation *candidateSliceOp) {
+  if (auto sliceOp = dyn_cast<tensor::InsertSliceOp>(candidateSliceOp)) {
+    return tileAndFuseConsumerOfSliceSCFFor(rewriter, sliceOp);
+  } else if (auto sliceOp =
+                 dyn_cast<tensor::ParallelInsertSliceOp>(candidateSliceOp)) {
+    return tileAndFuseConsumerOfSliceSCFForall(rewriter, sliceOp);
+  } else {
+    return failure();
+  }
+}
+
+//===----------------------------------------------------------------------===//
+//                  END OF FUSE CONSUMER COPY - AVARMA
+//===----------------------------------------------------------------------===//
+namespace {
+
+class AMDAIEFuseConsumerIntoLoopPass
+    : public impl::AMDAIEFuseConsumerIntoLoopBase<
+          AMDAIEFuseConsumerIntoLoopPass> {
+ public:
+  AMDAIEFuseConsumerIntoLoopPass() = default;
+  AMDAIEFuseConsumerIntoLoopPass(const AMDAIEFuseConsumerIntoLoopPass &pass) {}
+  AMDAIEFuseConsumerIntoLoopPass(
+      const AMDAIEFuseConsumerIntoLoopOptions &options)
+      : AMDAIEFuseConsumerIntoLoopBase(options) {}
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<scf::SCFDialect>();
+  }
+  void runOnOperation() override;
+};
+
+void AMDAIEFuseConsumerIntoLoopPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  mlir::FunctionOpInterface funcOp = getOperation();
+  IRRewriter rewriter(context);
+
+  // Walk through the graph in post order and find the loop.
+  Operation *scfLoopOp = nullptr;
+  funcOp->walk<WalkOrder::PostOrder, ReverseIterator>(
+      [&](LoopLikeOpInterface op) {
+        if (isa<scf::ForOp>(op) && useSCFFor) {
+          scfLoopOp = op;
+          return WalkResult::interrupt();
+        } else if (isa<scf::ForallOp>(op) && !useSCFFor) {
+          scfLoopOp = op;
+          return WalkResult::interrupt();
+        }
+        return WalkResult::advance();
+      });
+
+  if (!scfLoopOp) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "There is no scf.for/forall loop to fuse with.\n");
+    return;
+  }
+
+  LoopLikeOpInterface loops = cast<LoopLikeOpInterface>(scfLoopOp);
+  auto loopBody = useSCFFor ? cast<scf::ForOp>(loops).getBody()
+                            : cast<scf::ForallOp>(loops).getBody();
+
+  // Search the compute op and its consumer slices.
+  linalg::LinalgOp linalgOp;
+  loopBody->walk<WalkOrder::PostOrder, ReverseIterator>(
+      [&](linalg::LinalgOp op) {
+        if (linalg::isaContractionOpInterface(op)) {
+          linalgOp = op;
+          return WalkResult::interrupt();
+        }
+        return WalkResult::advance();
+      });
+
+  if (!linalgOp) {
```

Comment:
Here you should add another check to see if there's a following elementwise op after this contractionOp. If there's no such following op, then directly return. Otherwise it will run this pass and fail when there's matmul only.

Reply:
Very good point!

Well, the fuse-consumer API will take care of that check and return failure.

So on the caller site, instead of re-performing the same check, I have added an `emitWarning` to bail out cleanly.

Reply:
I've also added a test case to demonstrate the same.

Reply:
If there's an unpack op after forall loop, it will try to fuse unpack inside the loop. This needs to be avoided.

Reply:
@Abhishek-Varma Since you did force-push, it's hard for me to find the changes in your pass. Can you summarize how you solve the above problem?

Reply:
To unblock us - I've restricted the fusion to only those `Linalg` ops which are implementing both `TilingInterface` and `DestinationPassingStyle`.